cmd/compute-domain-kubelet-plugin/nvlib.go

```diff
@@ -76,11 +78,29 @@ func newDeviceLib(driverRoot root) (*deviceLib, error) {
 		nvml.WithLibraryPath(driverLibraryPath),
 	)
 	d := deviceLib{
-		Interface:         nvdev.New(nvmllib),
-		nvmllib:           nvmllib,
-		driverLibraryPath: driverLibraryPath,
-		devRoot:           driverRoot.getDevRoot(),
-		nvidiaSMIPath:     nvidiaSMIPath,
+		Interface:             nvdev.New(nvmllib),
+		nvmllib:               nvmllib,
+		driverLibraryPath:     driverLibraryPath,
+		devRoot:               driverRoot.getDevRoot(),
+		nvidiaSMIPath:         nvidiaSMIPath,
+		maxImexChannelCount:   0,
+		nvCapImexChanDevInfos: make([]*nvcapDeviceInfo, 0),
+	}
+
+	mic, err := d.getImexChannelCount()
+	if err != nil {
+		return nil, fmt.Errorf("error getting max IMEX channel count: %w", err)
+	}
+	d.maxImexChannelCount = mic
+
+	// Iterate through [0, mic-1] to pre-compute objects for CDI specs
+	// (major/minor dev node numbers won't change at runtime).
+	for i := range mic {
+		info, err := d.getNVCapIMEXChannelDeviceInfo(i)
+		if err != nil {
+			return nil, fmt.Errorf("error getting nvcap for IMEX channel '%d': %w", i, err)
+		}
+		d.nvCapImexChanDevInfos = append(d.nvCapImexChanDevInfos, info)
```

Comment:
The alternative here would be to just cache the major number of the nvcaps device so that `getNVCapIMEXChannelDeviceInfo()` no longer needed to return an error (but still would need to be called a bunch of times in the other function). I don't have a preference either way.

Reply:
I think I like this current way, the complete pre-computation seems to make sense and this looks neat on the consumer side.